app/services/puppet_class_importer.rb

```diff
@@ -171,7 +171,28 @@ def db_environments
   end
 
   def actual_environments
-    @proxy_envs ||= (proxy.environments.map(&:to_s) - ignored_environments)
+    (proxy_environments & (current_user_environments + to_be_created_environments)) - ignored_environments
+  end
+
+  def proxy_environments
+    proxy.environments.map(&:to_s)
+  end
+
+  def to_be_created_environments
+    proxy_environments - Environment.where(:name => proxy_environments).pluck(:name)
+  end
+
+  def current_user_taxonomy_ids
```

Comment:
I would say this and the method below really belong in the user model, this logic only related to fetching information from User:

```
class User
  def taxonomy_ids
    { :organizations => my_organizations.pluck(:id), :locations => my_locations.pluck(:id) }
  end
 
  def visible_environments
        Environment.unscoped
               .authorized(:view_environemnts)
               .joins(:taxable_taxonomies)
               .where('taxable_taxonomies.taxonomy_id' => taxonomy_ids[:organizations] + taxonomy_ids[:locations])
               .distinct
               .pluck(:name)
  end
end
```